src/mame/brother/lw350.cpp

```diff
@@ -0,0 +1,1749 @@
+		map(0x0, 0x0).r(FUNC(upd765a_device::msr_r));
+		map(0x1, 0x1).rw(FUNC(upd765a_device::fifo_r), FUNC(upd765a_device::fifo_w));
+	}
+
+	auto irq_wr_cb() { return intrq_wr_callback(); }
+	auto dreq_wr_cb() { return drq_wr_callback(); }
+	//auto dend_rd_cb() { return dend_cb.bind(); }
+};
+
+DEFINE_DEVICE_TYPE(HD63266F, hd63266f_t, "hd63266f", "Hitachi HD63266F FDC")
+
+hd63266f_t::hd63266f_t(const machine_config& mconfig, const char* tag, device_t* owner, uint32_t clock)
+: upd765_family_device(mconfig, HD63266F, tag, owner, clock) {
+	set_ready_line_connected(false);
+	set_select_lines_connected(false);
+	has_dor = false;
+}
+
+#include "imagedev/floppy.h"
+
+using lw350_floppy_connector = floppy_connector;
+#define LW350_FLOPPY_CONNECTOR FLOPPY_CONNECTOR
+#define LW350_FLOPPY FLOPPY_35_HD
+
+#endif
+
+class lw350_state : public driver_device
+{
+public:
+	lw350_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag),
+		maincpu(*this, "maincpu"),
+		screen(*this, "screen"),
+		floppy(*this, "floppy"),
+		fdc(*this, "fdc"),
+		beeper(*this, "beeper"),
+		io_kbrow(*this, "kbrow.%u", 0)
+	{ }
+
+	void lw350(machine_config& config);
+
+private:
+	// devices
+	required_device<hd64180rp_device> maincpu;
+	required_device<screen_device> screen;
+	required_device<lw350_floppy_connector> floppy;
+	required_device<hd63266f_t> fdc;
+	required_device<beep_device> beeper;
+	optional_ioport_array<9> io_kbrow;
+
+	uint8_t vram[80 * 128];
+	uint8_t io_70, io_7a, io_b8, io_90;
+
+	// screen updates
+	uint32_t screen_update(screen_device& screen, bitmap_rgb32& bitmap, const rectangle& cliprect);
+
+	uint8_t illegal_r(offs_t offset, uint8_t mem_mask = ~0) {
+		logerror("%s: unmapped memory read from %0*X & %0*X\n", machine().describe_context(), 6, offset, 2, mem_mask);
+		return 0;
+	}
+	void illegal_w(offs_t offset, uint8_t data, uint8_t mem_mask = ~0) {
+		logerror("%s: unmapped memory write to %0*X = %0*X & %0*X\n", machine().describe_context(), 6, offset, 2, data, 2, mem_mask);
+	}
+
+	// IO
+	void io_70_w(uint8_t data) {
+		io_70 = data;
+	}
+	uint8_t io_74_r() {
+		// 0x00: 7 lines display (64 pixels height)
+		// 0x80: 14 lines display (128 pixels height)
+		return 0x80;
+	}
+	uint8_t io_a8_r() {
+		// bit 0: case open
+		// bit 2: carriage return indicator
+		//return 0x01; // case open
+		return 0x00;
+	}
+	uint8_t io_b8_r() {
+		// keyboard matrix
+ 		if(io_b8 <= 8)
+			return io_kbrow[io_b8].read_safe(0);
+
+		switch(io_b8) {
+		// get language
+		case 0x09:  // valid results (see get_index_from_language)
+			//return ~0x20; // french
+			//return ~0x10; // french
+			//return ~0x08; // german
+			//return ~0x04; // french
+			//return ~0x02; // french
+			//return ~0x01; // german
+			return ~0x00; // german
+		default:   return 0x00;
+		}
+	}
+	void io_b8_w(uint8_t data) {
+		io_b8 = data;
+	}
+	void rombank_w(uint8_t data) { // E0
+		membank("rom")->set_entry(data & 0x03);
+	}
```

Comment:
Please use `required_memory_bank` to look these up so you don’t need to do tag lookups on accesses.